src/engine/enginemaster.cpp

```diff
@@ -366,16 +406,80 @@ void EngineMaster::process(const int iBufferSize) {
     // Mix all the PFL enabled channels together.
     m_headphoneGain.setGain(chead_gain);
 
-    if (m_bRampingGain) {
-        ChannelMixer::mixChannelsRamping(
-                m_headphoneGain, &m_activeHeadphoneChannels,
-                &m_channelHeadphoneGainCache,
-                m_pHead, iBufferSize);
+    // Mixxx gets input signals delayed by the input latency,
+    // so to preserve the relative timing of the input with the master and headphones
+    // outputs, delay the master, booth, and headphones outputs by the input latency.
+    // This way, if the microphone user is on beat with the output they hear coming from
+    // the speakers/headphones, the input will remain on beat at the expense of
+    // adding some latency.
+    // If the input signal is being directly monitored (mixed with the master output
+    // in hardware), do not delay the master output. The recording/broadcasting signal
+    // will be delayed in this case to keep the recorded/broadcasted signal on beat
+    // and consistent with what the user hears out the speakers/headphones.
+    // If no microphone inputs are configured, avoid adding this extra latency.
+    TalkoverMixMode configuredTalkoverMixMode = static_cast<TalkoverMixMode>(
+        static_cast<int>(m_pTalkoverMixMode->get()));
+    // These latency values are in milliseconds as required by EngineDelay.
+    double inputLatencyCompensation;
+    double measuredRoundTripLatency = m_pRoundTripLatency->get();
```

Comment:
> I disagree. This is way too complicated to be exposed to users and users do not need to understand it. All they need to do is measure the real roundtrip latency with a loopback cable and Mixxx should do the math correctly.

Yes, Mixxx should do this, but not in a CO interface hiding the truth. It should calculate the pure delay values in a preferences pane and not in the engine. That is my point.